src/CattyTests/DataModel/UserVariableTests.swift

```diff
@@ -70,4 +73,91 @@ final class UserVariableTests: XCTestCase {
         XCTAssertFalse(userVariableA === userVariableCopy)
         XCTAssertTrue(userVariableB === userVariableCopy)
     }
+
+    func testIsEqualToUserVariableForDirectAssignment() {
+        let userVariableA = UserVariable()
+        userVariableA.name = "userVar"
+        userVariableA.value = self.arrayWithSameElement
+
+        var userVariableB = userVariableA
+        XCTAssertTrue(userVariableA.isEqual(to: userVariableB))
+        XCTAssertEqual(userVariableB.value as! NSMutableArray, self.arrayWithSameElement)
+
+        userVariableA.value = self.arrayWithDifferentElement
+        XCTAssertTrue(userVariableA.isEqual(to: userVariableB))
+        XCTAssertEqual(userVariableB.value as! NSMutableArray, self.arrayWithDifferentElement)
+
+        let userVariableC = UserVariable()
+        userVariableC.name = "userVarC"
+
+        userVariableB = userVariableC
+        XCTAssertFalse(userVariableA.isEqual(to: userVariableB))
+    }
+
+    func testIsEqualToUserVariableForInitWithNameAndIsList() {
+        let userVariableA = UserVariable()
+        userVariableA.name = "userVar"
+        userVariableA.value = self.arrayWithSameElement
+
+        var userVariableB = UserVariable(name: userVariableA.name, isList: userVariableA.isList)
+        XCTAssertFalse(userVariableB!.isEqual(to: userVariableA))
+        XCTAssertNil(userVariableB?.value)
+
+        userVariableB?.value = userVariableA.value
+        XCTAssertTrue(userVariableB!.isEqual(to: userVariableA))
+        XCTAssertEqual(userVariableB?.value as! NSMutableArray, self.arrayWithSameElement)
+
+        userVariableA.value = self.arrayWithDifferentElement
+        userVariableB?.value = userVariableA.value
+        XCTAssertTrue(userVariableA.isEqual(to: userVariableB))
+        XCTAssertEqual(userVariableB?.value as! NSMutableArray, self.arrayWithDifferentElement)
+
+        userVariableB = UserVariable(name: userVariableA.name, isList: !userVariableA.isList)
+        XCTAssertFalse(userVariableB!.isEqual(to: userVariableA))
+    }
+
+    func testIsEqualToUserVariableForInitWithUserVariable() {
+        let userVariableA = UserVariable()
+        userVariableA.name = "userVar"
+        userVariableA.value = self.arrayWithSameElement
+
+        var userVariableB = UserVariable(variable: userVariableA)
```

Comment:
This initialization is alright, but it makes code review really hard, because I needed to look at this initializer since it was not obvious whether the `value` is also getting copied or not. Recommendation: always initialize the actual and expected instances the same way (i.e. using `UserVariable(name: ,isList: )`).

Reply:
Hi, @m-herold I used this initializer to create a new `userVariable` using the existing variable and test. But  If I use the `UserVariable(name: ,isList: )` initializer then there would be no difference between this test and the test before it. So shall I remove this test or look for a different approach:

```
let userVariable = UserVariable(name: "userVar", isList: false )
let userVarA = UserVariable(variable: userVariable)
let userVarB = UserVariable(variable: userVariableA)
userVariableA?.value = NSMutableArray(array: [50, 51])
userVariableB?.value = NSMutableArray(array: [50, 52])
XCTAssertFalse(userVariableB!.isEqual(to: userVariableA))

userVariableB?.value = NSMutableArray(array: [50, 51])
XCTAssertTrue(userVariableB!.isEqual(to: userVariableA))
```

I am not sure if this approach is correct or not.

Reply:
You approach looks almost good, except of the initializer, which is again not intuitive. I would recommend something like:
```
let userVariableA = UserVariable(name: "userVar", isList: true)
let userVariableB = UserVariable(name: "userVar", isList: true)

userVariableA?.value = NSMutableArray(array: [50, 51])
userVariableB?.value = NSMutableArray(array: [50, 52])
XCTAssertFalse(userVariableB!.isEqual(to: userVariableA))

userVariableB?.value = NSMutableArray(array: [50, 51])
XCTAssertTrue(userVariableB!.isEqual(to: userVariableA))
```

This would be a good test for verifying one code executing path of the `isEqual` method (namely for verifying the values). I would then add additional tests doing the same mutation for `name` and `isList`.

> But If I use the UserVariable(name: ,isList: ) initializer then there would be no difference between this test and the test before it. 

I am not sure what you mean by that, but as far as I remember there was an assignment like `var userVariableB = userVariableA`. Note: this is not the same, although the variables have the same values. This is not a new initialization, thus they share the same address in memory hence the excepted and the actual values are always the same, that would be something similar to pseudo code `assert that a is equal to a` (which is a tautology).